### PR TITLE
Pull in upstream updates from cisagov/skeleton-tf-module

### PR DESCRIPTION
**Pull request by jsf9k**
## 🗣 Description

This pull request pulls in upstream changes from [cisagov/skeleton-tf-module](https://github.com/cisagov/skeleton-tf-module).

## 💭 Motivation and Context

I need this module to pass the "Terraform validate" pre-commit hook so that cisagov/ansible-role-openvpn#15 can pass.

## 🧪 Testing

_TO DO_

## 🚥 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: providers.tf
Comment (jsf9k) [inline providers.tf]: Thanks for adding this commit @mcdonnnj!

Do we prefer this way, or do we want to define a default region variable with a default value of `us-east-1`?  It doesn't matter very much, but whatever we do here is likely to become the standard for all our Terraform modules.  We should get it the way we want it.

I personally think it makes sense to create the variable, on the off chance that someone wants to use this module in a single-account setup where all providers are really just the default provider.

Comment (mcdonnnj) [reply providers.tf]: I originally had a variable in my testing branch. I switched to the non-variable way to follow how the module is used in the example. I am happy to switch to a variable with a default value if we want to go in that direction.

Comment (dav3r) [reply providers.tf]: When in doubt, make it a variable.  I'd rather err on the side of flexibility.

Comment (mcdonnnj) [reply providers.tf]: @dav3r I felt the same way and made this change in ad358f1

File: README.md
Comment (dav3r) [inline README.md]: I don't think this single account example is correct.  Unless I'm mistaken, I think it still needs a providers block, but it would look like this:
```
providers = {
  aws                   = aws
  aws.images-production = aws
  aws.images-staging    = aws
}
```

Also, I think showing that we can override the default value of the `aws_region` variable here might be confusing, since we don't show it in the "multi-account" example.  When I look at this, it implies that doing this has something to do with the fact that this is a single-account example, even though it does not.

**Review by dav3r — CHANGES_REQUESTED** [review]
Everything looks good, except for the new single vs. multi-account examples.  Please take a look at my comments.

File: README.md
Comment (mcdonnnj) [reply README.md]: If we're going to expect using the module to include passing the providers explicitly, then I don't think we should have a module variable for the AWS region. The variable only serves a purpose if we are modifying the providers the module uses by default. I think having a variable that does nothing if following our examples is equally confusing. We could switch to a local variable in `providers.tf` to ensure that the default is consistent. @jsf9k what are your thoughts?

**Review by felddy — APPROVED** [review]
Haha... Wow. @jsf9k was just looking for a little **validation**  but apparently came up SHORT! 🍿

File: README.md
Comment (dav3r) [reply README.md]: Here's what is confusing me- and maybe I'm misunderstanding something here.  If we allow the providers to be inherited from the module caller, like so:
```
module "example" {
  source = "github.com/cisagov/molecule-iam-user-tf-module"

  aws_region = "us-east-2"

  ssm_parameters = ["/example/parameter1", "/example/config/*"]
  user_name      = "test-molecule-iam-user-tf-module"

  tags = {
    Team        = "VM Fusion - Development"
    Application = "molecule-iam-user-tf-module testing"
  }
}
```

Then those providers still need to be defined somewhere by the caller, which looks something like this (lifted from the example):
```
# Default AWS provider (ProvisionAccount for the Users account)
provider "aws" {
  region  = "us-east-1"
  profile = "cool-users-provisionaccount"
}

# ProvisionParameterStoreReadRoles AWS provider for the
# Images (Production) account
provider "aws" {
  region  = "us-east-1"
  profile = "cool-images-production-provisionparameterstorereadroles"
  alias   = "images-production"
}

# ProvisionParameterStoreReadRoles AWS provider for the
# Images (Staging) account
provider "aws" {
  region  = "us-east-1"
  profile = "cool-images-staging-provisionparameterstorereadroles"
  alias   = "images-staging"
}
```

The `region` is required in those providers defined by the caller.  So even if we define a variable for the region and pass that into the module, the module is still going to inherit the providers (and corresponding regions) defined by the caller, right?

If all that is correct, then it means that there is no point in passing a variable into the module for the region.  If something about that is wrong, please let me know where.

BTW, I wanted to test this out locally to verify if the module's provider region (i.e. the one set by the variable) can override the provider region inherited from the caller, but I don't think I can in this case since all of the resources that get created are global (IAM user and roles)- I would have no way of knowing which region was actually used.

Comment (jsf9k) [reply README.md]: This all started because these Terraform modules are starting to fail the terraform-validate pre-commit hook.  It sounds like we just need to put _something_ there for the region in order to get that hook to pass, and that something will always be ignored.  I'm fine with that being a local variable, or not a variable at all.

@mcdonnnj, do you want to revert ad358f1?

Comment (mcdonnnj) [reply README.md]: So this is a more complete discussion let me background what's happening. These modules are not starting to fail the `terraform validate` pre-commit hook. In June of last year (2019) the `pre-commit-terraform` hook did some Terraform v0.12 updates. As seen [on this line](https://github.com/antonbabenko/pre-commit-terraform/commit/35e0356188b64a4c5af9a4e7200d936e514cba71#diff-0560dc7f5fcce3201df9ddfd7eef7d1fL37), the hook we were using, `terraform_validate_no_variables`, was combined and renamed to `terraform_validate`. It wasn't  caught until last month (March of 2020), and the hook ID being used was fixed in https://github.com/cisagov/skeleton-generic/pull/40. So I believe this would have been failing for us ever since Terraform v0.12 changes, we just didn't know because of a misconfiguration.

@dav3r 
Reading from [Terraform documentation](https://www.terraform.io/docs/configuration/modules.html#passing-providers-explicitly) it looks like the way you provided is how we should show (and do) single account usage. I haven't really seen much alias usage before so I was not well read on the particulars that introduces.

I have two possibilities in mind to resolve this:
1. We use a local variable in `providers.tf` to set the region and thus it will pass, even if the region is somewhat superfluous.
2. Everyone can make sure they have an `AWS_DEFAULT_REGION` environment value set. Terraform will search for that variable as part of resolving the `region` for the AWS provider as described:
    > `region` - (Optional) This is the AWS region. It must be provided, but it can also be sourced from the `AWS_DEFAULT_REGION` environment variables, or via a shared credentials file if `profile` is specified.

    A corresponding variable would be configured for the GitHub Actions workflow to provide the same functionality (I have tested this locally after removing the `region` declarations in `providers.tf` for the module).

I'm leaning toward the environment variable because that would be a blanket fix for this in the future instead of continuously needing to provide explicitly declared `region` arguments to the AWS provider.

@jsf9k I'd probably just do a fresh commit to resolve the variable situation. I would like to keep the single account example and just update it to reflect the usage @dav3r showed.

Edited to clean up some formatting to more closely match the source material.

Comment (felddy) [reply README.md]: Thanks for the great explanation.  I'm not sure if I'm understanding that quote correctly:
> ...or via a shared credentials file if profile is specified.

Is this referring to the `~/.aws/config` file:
```
[default]
region = us-east-1
```

Comment (mcdonnnj) [reply README.md]: Along those lines. Sorry I missed that in my hunt for sources for clear explanation. That option is explicitly laid out in the [AWS provider reference](https://www.terraform.io/docs/providers/aws/index.html).
> `shared_credentials_file` = (Optional) This is the path to the shared credentials file. If this is not set and a profile is specified, `~/.aws/credentials` will be used.

So it will only do the search if a `profile` is declared for the provider.

Comment (mcdonnnj) [reply README.md]: After a team discussion we opted for the local variable option from the two options I brought forward. This can be seen implemented in 326483f.

**Review by dav3r — APPROVED** [review]
It was a long road, but I like where we ended up! 🌻

**Comment by jsf9k** [comment]
@mcdonnnj, go ahead and merge this when you think it is good.  I haven't been following this one as closely as you have, so I will defer to your good judgement.

**Review by mcdonnnj — APPROVED** [review]
I see no issues so I'm going to go ahead and merge this PR.

**Pull request merged**
